### PR TITLE
Stop elaborating errors when relating intersection constituents

### LINKED ISSUE #7199: String literal types in complex type expressions
**TypeScript Version:** 

1.8.2

**Code**

```ts
interface Foo {
    ok:string;
    okToo: number;
    notOk:  "o" | "k";
}

interface Bar {
    bar: string;
}

interface FooBar extends Foo, Bar{
}

function mix<T>(obj:T): T & Bar {
    (obj as any).bar="bar";
    return obj as T & Bar;
}

var fooBar:FooBar = mix({
    ok:"ok",
    okToo: 42,
    notOk:"k"
});
```

**Expected behavior:**

The assignment to `fooBar` should be allowed. 

**Actual behavior:** 

I get the totally confusing error:

```
Error:(20, 5) TS2322: Type '{ ok: string; notOk: string; } & Bar' is not assignable to type 'FooBar'.
 Type 'Bar' is not assignable to type 'FooBar'.
 Property 'ok' is missing in type 'Bar'.
```

The confusing part is that it complains about the first property (`ok`) and not about `notOk`.

Looking at the error message it seems that the string literal type `notOk:"o"|"k"` was converted to `notOk :string`.

Therefore when I turn the string literal type to a string (`notOk:string`) everything is OK....

Hint: defining the string literal type as a type does *not* solve the problem:
```ts
type NotOK =  "o" | "k";

interface Foo {
    //...
    notOk: NotOK;
}
//...
```

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #7199

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5522,7 +5522,7 @@ namespace ts {
                     // A & B = (A & B) | (C & D).
                     if (source.flags & TypeFlags.Intersection) {
                         // If target is a union type the following check will report errors so we suppress them here
-                        if (result = someTypeRelatedToType(<IntersectionType>source, target, reportErrors && !(target.flags & TypeFlags.Union))) {
+                        if (result = someTypeRelatedToType(<IntersectionType>source, target, reportErrors && !(target.flags & (TypeFlags.Union | TypeFlags.ObjectType)))) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: Seems like we should use structural error reporting for other things, too, like primitives or type parameters.

**Comment by DanielRosenwasser** [comment]
@sandersn @RyanCavanaugh any thoughts?

**Comment by RyanCavanaugh** [comment]
Seems like the change is simple enough and the baselines look improved. What's the worst-case scenario here?

**Comment by sandersn** [comment]
The worst scenario I could come up with is the case that the error message happens when relating a multiple intersection of giant types, say 1000+ members each. Then the error "A is not assignable to B & C & D & E & F & G -- 'A.a' is 'string' but 'B & C & D & E & F & G.a' is 'number'." means you have to go searching through B, C, D, E, F and G to find and fix property 'a'.

This kind of type is common in big data/relational code, say [Spark](http://spark.apache.org/) or [U-SQL](https://msdn.microsoft.com/en-us/library/azure/mt591959.aspx). In U-SQL (which I'm more familiar with), the above type results from joining giant tables, eg `Clicks JOIN Pages JOIN Users JOIN Ads`. Right now Spark officially supports Scala, Python and R but [there is already an unofficial Node project](https://github.com/henridf/apache-spark-node). So real users could run into this scenario someday.

On the other hand, I think insufficient elaboration is better than confusingly-wrong elaboration, so I vote that we take this change. :+1:

**Comment by RyanCavanaugh** [comment]
:+1:

**Pull request merged**